File: android/app/src/main/java/com/trashed/driver/NativeDockAvatarPolicy.java
package com.trashed.driver;

import java.net.URI;
import java.util.Locale;

/** Pure allowlist for saved profile images. Never an authority or cookie policy. */
final class NativeDockAvatarPolicy {
    private NativeDockAvatarPolicy() { }

    static String imageURL(String raw, String origin) {
        if (raw == null || origin == null || raw.isEmpty() || raw.length() > 2048 || forbiddenCharacters(raw)) return null;
        try {
            URI base = new URI(origin);
            if (!validOrigin(base)) return null;
            URI value = new URI(raw);
            if (value.getRawUserInfo() != null || value.getRawFragment() != null) return null;
            boolean relative = !value.isAbsolute();
            if (relative && (!raw.startsWith("/") || raw.startsWith("//") || value.getRawAuthority() != null)) return null;
            // Inspect before resolve(): URI.resolve normalizes dot segments.
            if (!safePath(value.getRawPath())) return null;
            URI url = relative ? base.resolve(value) : value;
            String host = url.getHost();
            if (host == null || url.getRawUserInfo() != null || url.getRawFragment() != null) return null;
            host = host.toLowerCase(Locale.ROOT);
            if (!sameOrigin(url, base)) {
                if (!"https".equals(url.getScheme()) || (url.getPort() != -1 && url.getPort() != 443)) return null;
                boolean known = host.equals("api.dicebear.com") || host.equals("avatars.githubusercontent.com")
                    || host.matches("lh[0-9]+\\.googleusercontent\\.com")
                    || host.matches("[a-z0-9-]+\\.public\\.blob\\.vercel-storage\\.com");
                if (!known) return null;
            }
            String path = url.getRawPath();
            if (host.equals("api.dicebear.com") && path.matches("/[0-9]+\\.x/[a-z][a-z0-9-]*/svg")) {
                // Equivalent raster representation of the saved avatar. Preserve
                // seed, colors and every option byte-for-byte; do not render SVG.
                int queryAt = raw.indexOf('?');
                String query = queryAt < 0 ? "" : raw.substring(queryAt);
                return url.getScheme() + "://" + url.getRawAuthority() + path.substring(0, path.length()-3) + "png" + query;
            }
            return url.toASCIIString();
        } catch (Exception invalid) {
            return null;
        }
    }

    private static boolean validOrigin(URI origin) {
        if (origin.getHost() == null || origin.getRawUserInfo() != null || origin.getRawQuery() != null || origin.getRawFragment() != null
            || !(origin.getRawPath().isEmpty() || origin.getRawPath().equals("/"))) return false;
        String host = origin.getHost().toLowerCase(Locale.ROOT);
        boolean loopback = host.equals("127.0.0.1") || host.equals("localhost");
        // Cleartext is only for an explicitly configured local QA origin.
        if ("http".equals(origin.getScheme())) return loopback && origin.getPort() > 0;
        return "https".equals(origin.getScheme()) && !loopback && !host.contains(":") && !host.matches("[0-9.]+")
            && !host.endsWith(".localhost") && (origin.getPort() == -1 || origin.getPort() > 0);
    }

    private static boolean sameOrigin(URI a, URI b) {
        return a.getScheme().equals(b.getScheme()) && a.getHost().equalsIgnoreCase(b.getHost()) && port(a) == port(b);
    }
    private static int port(URI value) {
        return value.getPort() == -1 ? ("https".equals(value.getScheme()) ? 443 : 80) : value.getPort();
    }
    private static boolean forbiddenCharacters(String value) {
        for (int i = 0; i < value.length(); i++) {
            char c = value.charAt(i);
            if (c <= 32 || c == 127 || c == '\\') return true;
        }
        return false;
    }
    private static boolean safePath(String path) {
        if (path == null || path.isEmpty() || !path.startsWith("/")) return false;
        // Encoded separators/control/dot/percent sequences have no legitimate
        // role in the supported avatar paths. Reject instead of normalizing.
        if (path.toLowerCase(Locale.ROOT).matches(".*%(?:00|0a|0d|2e|2f|5c|25).*")) return false;
        for (String piece : path.split("/")) if (piece.equals(".") || piece.equals("..")) return false;
        return true;
    }

    static String initials(String name) {
        if (name == null) return "?";
        String cleaned = name.replaceAll("[\\p{Cc}\\p{Cf}]", "").trim();
        String[] words = cleaned.split("[\\s\\p{Z}]+");
        if (cleaned.isEmpty() || words.length == 0) return "?";
        String first = new String(Character.toChars(words[0].codePointAt(0)));
        String last = words.length > 1 ? new String(Character.toChars(words[words.length-1].codePointAt(0))) : "";
        String upper = (first + last).toUpperCase(Locale.ROOT);
        int count = upper.codePointCount(0, upper.length());
        return count > 2 ? upper.substring(0, upper.offsetByCodePoints(0, 2)) : upper;
    }
}

File: ios/App/App/NativeDockAvatarPolicy.swift
import Foundation

/// Saved-profile image allowlist. This policy does not confer account authority.
enum NativeDockAvatarPolicy {
    /// Capacitor may include appStartPath in serverURL. Derive a trusted API
    /// origin here without relaxing validation of saved image URL values.
    static func origin(fromConfiguredURL url: URL) -> URL? {
        guard var parts = URLComponents(url: url, resolvingAgainstBaseURL: false),
              parts.user == nil, parts.password == nil else { return nil }
        parts.path = ""; parts.query = nil; parts.fragment = nil
        return validOrigin(parts) ? parts.url : nil
    }

    static func imageURL(_ raw: String?, origin: URL) -> URL? {
        guard let raw = raw, !raw.isEmpty, raw.utf16.count <= 2048,
              !raw.unicodeScalars.contains(where: { $0.value <= 32 || $0.value == 127 || $0 == "\\" }),
              let base = URLComponents(url: origin, resolvingAgainstBaseURL: false), validOrigin(base),
              // Foundation may percent-encode malformed input; reject rather than repair.
              raw.range(of: "%($|[^0-9A-Fa-f]|[0-9A-Fa-f]($|[^0-9A-Fa-f]))", options: .regularExpression) == nil,
              let value = URLComponents(string: raw), value.user == nil, value.password == nil, value.fragment == nil,
              safePath(value.percentEncodedPath) else { return nil }
        let url: URL
        if value.scheme == nil {
            guard raw.hasPrefix("/"), !raw.hasPrefix("//"), value.host == nil,
                  let resolved = URL(string: raw, relativeTo: origin)?.absoluteURL else { return nil }
            url = resolved
        } else {
            guard let absolute = value.url else { return nil }
            url = absolute
        }
        guard var parts = URLComponents(url: url, resolvingAgainstBaseURL: false),
              let host = parts.host?.lowercased(), parts.user == nil, parts.password == nil, parts.fragment == nil else { return nil }
        if !sameOrigin(parts, base) {
            guard parts.scheme == "https", parts.port == nil || parts.port == 443 else { return nil }
            let known = host == "api.dicebear.com" || host == "avatars.githubusercontent.com"
                || host.range(of: "^lh[0-9]+\\.googleusercontent\\.com$", options: .regularExpression) != nil
                || host.range(of: "^[a-z0-9-]+\\.public\\.blob\\.vercel-storage\\.com$", options: .regularExpression) != nil
            guard known else { return nil }
        }
        if host == "api.dicebear.com",
           parts.percentEncodedPath.range(of: "^/[0-9]+\\.x/[a-z][a-z0-9-]*/svg$", options: .regularExpression) != nil {
            // Raster equivalent of the exact saved selection; preserve all query
            // bytes, seed and style options rather than rendering arbitrary SVG.
            parts.percentEncodedPath = String(parts.percentEncodedPath.dropLast(3)) + "png"
        }
        return parts.url
    }

    private static func validOrigin(_ value: URLComponents) -> Bool {
        guard let host = value.host?.lowercased(), value.user == nil, value.password == nil,
              value.query == nil, value.fragment == nil, value.path.isEmpty || value.path == "/" else { return false }
        let loopback = host == "127.0.0.1" || host == "localhost"
        if value.scheme == "http" { return loopback && (value.port ?? 0) > 0 }
        return value.scheme == "https" && !loopback && !host.contains(":") && !host.hasSuffix(".localhost")
            && host.range(of: "^[0-9.]+$", options: .regularExpression) == nil
            && (value.port == nil || value.port! > 0)
    }
    private static func sameOrigin(_ left: URLComponents, _ right: URLComponents) -> Bool {
        left.scheme == right.scheme && left.host?.lowercased() == right.host?.lowercased()
            && (left.port ?? (left.scheme == "https" ? 443 : 80)) == (right.port ?? (right.scheme == "https" ? 443 : 80))
    }
    private static func safePath(_ path: String) -> Bool {
        guard path.hasPrefix("/"), path.range(of: "%(00|0a|0d|2e|2f|5c|25)", options: [.regularExpression, .caseInsensitive]) == nil else { return false }
        return !path.split(separator: "/").contains(where: { $0 == "." || $0 == ".." })
    }
    static func initials(_ name: String?) -> String {
        guard let name = name else { return "?" }
        let cleaned = name.unicodeScalars.filter { !CharacterSet.controlCharacters.contains($0) && !CharacterSet.illegalCharacters.contains($0) }
        let words = String(String.UnicodeScalarView(cleaned)).split(whereSeparator: { $0.isWhitespace })
        guard let first = words.first?.first else { return "?" }
        let last = words.count > 1 ? String(words.last!.first!) : ""
        return String((String(first) + last).uppercased().prefix(2))
    }
}
